Cf create services in multiclouddeploy step

fwilhe opened the pull request:
# Changes

This PR brings the feature to create cf services which the SAP Cloud SDK Pipeline already has to the multiclouddeploy.

- [x] Tests
- [x] Documentation

daniel-kurzynski commented on test/groovy/MulticloudDeployTest.groovy:
```diff
@@ -95,6 +87,9 @@ class MulticloudDeployTest extends BasePiperTest {
             ]
         ]
+
+        helper.registerAllowedMethod('echo', [CharSequence.class], {s -> println(s)})
```
Debug output?

fwilhe replied:
No, what would you suggest? I think it is ok to print echos in a test.

marcusholl replied:
In my personal opinion there is no benefit when printing info during tests at all. Nobody will read that normally.  In case some log output (remember perform logging in groovy basically with `echo`) is important it should be asserted. I vote for either removing the `println`. When everybody does that we end up in a huge amount of test logs which are not helpful at all.

daniel-kurzynski commented on test/groovy/MulticloudDeployTest.groovy:
```diff
@@ -276,4 +271,52 @@ class MulticloudDeployTest extends BasePiperTest {
+            source: 'file.mtar'
+        ])
+
+        assert closureRun != null
```
Instead of just asserting the parrellel execution, I am missing a test which checks for the service creation itself. E.g. that cfCreateService is called.

marcusholl replied:
dito -> `test/groovy/util/JenkinsMockStepRule.groovy`

fwilhe replied:
You're right, I've added an additional test case for that.

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -44,7 +47,8 @@ void call(parameters = [:]) {
-            .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS)
+            .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS, CONFIG_KEY_COMPATIBILITY)
+            .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS, CONFIG_KEY_COMPATIBILITY)
             .mixin(parameters, PARAMETER_KEYS)
```
Could you also please merge in the stage configuration. We will need that later on. Maybe even now. 
It would be create if you could remove passing stage configuration as parameter to this step but letting the step resolve the stage config instead. 

https://github.com/SAP/cloud-s4-sdk-pipeline-lib/blob/master/vars/stageProductionDeployment.groovy#L60
Would be great to get rid of passing the parameters cfTargets, neoTargets, cfCreateServices. 
But can also be done in a follow up.

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -13,23 +12,27 @@ import static com.sap.piper.Prerequisites.checkScript
+    'parallelExecution'
 ]
 
 @Field Set STEP_CONFIG_KEYS = []
```
```suggestion
@Field Set STEP_CONFIG_KEYS = GENERAL_CONFIG_KEYS
```

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -13,23 +12,27 @@ import static com.sap.piper.Prerequisites.checkScript
 
 @Field Set STEP_CONFIG_KEYS = []
 
 @Field Set PARAMETER_KEYS = GENERAL_CONFIG_KEYS.plus([
```
```suggestion
@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS.plus([
```

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -13,23 +12,27 @@ import static com.sap.piper.Prerequisites.checkScript
-    'source'
+    'source',
+    /** Defines Cloud Foundry service instances to create as part of the deployment.*/
+    'cfCreateServices'
```
I would move that at least to the step config.

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -13,23 +12,27 @@ import static com.sap.piper.Prerequisites.checkScript
     /** The source file to deploy to SAP Cloud Platform.*/
-    'source'
+    'source',
+    /** Defines Cloud Foundry service instances to create as part of the deployment.*/
```
You should maybe also explain which keys are supported in each instance of the list.

daniel-kurzynski replied:
How should a user know that he can define apiEndpoint, credentialsId, serviceManifest, manifestVariablesFiles,org,space ...

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -114,17 +139,21 @@ void call(parameters = [:]) {
-        } else {
-            echo "Executing deployments in sequence"
-            def closuresToRun = deployments.values().asList()
-            Collections.shuffle(closuresToRun) // Shuffle the list so no one tries to rely on the order of execution
```
No shuffling anymore?

marcusholl commented on vars/multicloudDeploy.groovy:
```diff
@@ -53,14 +57,35 @@ void call(parameters = [:]) {
+            def createServices = [:]
+            for (int i = 0; i < parameters.cfCreateServices.size(); i++) {
+                Map createServicesConfig = parameters.cfCreateServices[i]
+                createServices["Service Creation ${i + 1 > 1 ? i + 1 : ''}"] = {
```
* why not `i>0` instead of `i+1>1`?

This would end up in (for e.g. 3 services
```
- Service Creation
- Service Creation 2
- Service Creation 3
```

What would be wrong with having `Service Creation 1` for the first entry? Wouldn't it be more reasonable to drop the index in case there is only one entry (parameters.cfCreateServices.size())``

By the way: when we run in parallel (or introduce shuffling again)  the order of the items in random. Would end up in e.g. 

```
- Service Creation 3
- Service Creation 
- Service Creation 2
```

Looks strange and the index is more the less not helpful during troubleshooting. Can we have somehow a better name here (e.g. outlines where a service is created?

fwilhe replied:
honest answer: copy and paste from what we had so far in the sdk pipeline without too much considerations for such details. I'm ok with discussing how to improve it, but I'd prefer to keep it simple if possible.

marcusholl replied:
[meta]

> but I'd prefer to keep it simple if possible.

I guess all of us are always anxious to keep things simple. You will not find anybody fighting for the opposite. The question is what is simple? This question is not always easy to answer. The strive for simplicity must be seen in the light of some condition which we should fulfil. E.g. each feature comes with some additional code and some additional complexity. Now it would be easy to get simple code by not having any feature implemented. The resulting code would be simple, but useless ... A feature could also be: "make the life of a troubleshooter easy by having understandable log output". The log output makes the code more complex and less simple, but there is a benefit from that ...

marcusholl commented on vars/multicloudDeploy.groovy:
```diff
@@ -114,17 +139,21 @@ void call(parameters = [:]) {
+
+def runClosures(Map config, Map toRun, String label = "closures") {
+    echo "Executing $label"
+    if (config.parallelExecution) {
```
[curiosity] What is the reason for making parallelism configurable. Are there cases where we expect issue with parallel execution? By the way: I understand that this was configurable already before, from that point of view my question is not fully appropriate here in that PR ...

daniel-kurzynski replied:
It can break blue ocean. Blue ocean cannot display two levels of parallelity. So if you have a parrallel stage which has a parallel block the pipeline on blue ocean looks strange but it still works.

fwilhe replied:
Yes, I've checked if blue ocean has improved in those cases, but it has not. It looks really confusing, this is why we did not enable it by default.

marcusholl commented on vars/multicloudDeploy.groovy:
```diff
@@ -114,17 +139,21 @@ void call(parameters = [:]) {
+        parallel toRun
+    } else {
+        echo "Executing $label in sequence"
+        def closuresToRun = toRun.values().asList()
```
In case shuffling is important we can/should shuffle here.

fwilhe replied:
No it is not important, and in the sdk pipeline we removed in meanwhile, thus also removing it here

fwilhe commented:
@daniel-kurzynski @marcusholl 

Thanks for your reviews. I think I addressed all points that I consider mandatory.

Open as of now:

- Naming of the services keys https://github.com/SAP/jenkins-library/pull/1736#discussion_r448139682
- Getting rid of params https://github.com/SAP/jenkins-library/pull/1736#discussion_r447799167

I'm not opposed to changing anything there, but I'm not sure if we have a common understanding of how it should be. Do you consider those mandatory?

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -53,14 +68,35 @@ void call(parameters = [:]) {
+            def createServices = [:]
+            for (int i = 0; i < parameters.cfCreateServices.size(); i++) {
+                Map createServicesConfig = parameters.cfCreateServices[i]
+                createServices["Service Creation ${i + 1 > 1 ? i + 1 : ''}"] = {
```
```suggestion
                createServices["Service Creation ${i + 1}"] = {
```

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -53,14 +68,35 @@ void call(parameters = [:]) {
         def index = 1
         def deployments = [:]
 
+        if (parameters.cfCreateServices) {
```
```suggestion
        if (config .cfCreateServices) {
```

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -53,14 +68,35 @@ void call(parameters = [:]) {
 
+        if (parameters.cfCreateServices) {
+            def createServices = [:]
+            for (int i = 0; i < parameters.cfCreateServices.size(); i++) {
```
```suggestion
            for (int i = 0; i < config .cfCreateServices.size(); i++) {
```

daniel-kurzynski commented on vars/multicloudDeploy.groovy:
```diff
@@ -53,14 +68,35 @@ void call(parameters = [:]) {
+        if (parameters.cfCreateServices) {
+            def createServices = [:]
+            for (int i = 0; i < parameters.cfCreateServices.size(); i++) {
+                Map createServicesConfig = parameters.cfCreateServices[i]
```
```suggestion
                Map createServicesConfig = config .cfCreateServices[i]
```

daniel-kurzynski replied:
This place here is missing

fwilhe commented:
> see comments

copy and waste. sorry for missing that.

btw, I did not accept your suggestions because of the extra spaces, I made the changes manually and pushed.

daniel-kurzynski commented:
/it
